File: teensy_cs.py
import serial
import logging

from time import sleep



class Teensy_CS:
    '''
        Class to interface with the teensy_cs library
    '''

    CMD_SET_WIDTH   =   0x01
    CMD_SET_DELAY   =   0x02
    CMD_RUN	        =   0x11
    CMD_STOP        =	0x12
    CMD_RUN_TRIG    =   0x13
    CMD_PREPARE_TARGET  =   0x14
    CMD_CHECK       =   0x15

    ERR_TRIG_IN     =   0x80

    REPLY_LOG       =   0xa5


    def __init__(self, log = 0):
        self.teensy = serial.Serial('/dev/ttyACM0')
        self.log = log

        self.async_cmd = 0

    def set_pulse_width(self, w):
        ''' Sets the pulse width on the teensy '''
        self.teensy.write([self.CMD_SET_WIDTH])
        self.teensy.write(w.to_bytes(4, 'little'))
        _r = self.read(self.CMD_SET_WIDTH)

    def set_delay(self, d):
        ''' Sets the glitch delay on the teensy '''
        self.teensy.write([self.CMD_SET_DELAY])
        self.teensy.write(d.to_bytes(4, 'little'))
        _r = self.read(self.CMD_SET_DELAY)

    def run(self, trigger = 0, sync = 1):
        ''' Run the teensy code. Set trigger to 1 for the teensy to trigger the chipwhisperer '''
        _r = None
        if trigger:
            CMD = self.CMD_RUN_TRIG
        else:
            CMD = self.CMD_RUN
        self.teensy.write([CMD])
        if sync:
            _r = self.read(CMD)
        else:
            self.async_cmd = CMD
        return _r

    def read_async(self):
        ''' 
            Read the result of an asynchronous operation - called with self.run(sync = 0)
        '''
        return self.read(self.async_cmd)

    def check(self):
        self.teensy.write([self.CMD_CHECK])
        _r = self.read(self.CMD_CHECK)
        return _r


    def setup_target(self):
        self.teensy.write([self.CMD_PREPARE_TARGET])
        _r = self.read(self.CMD_PREPARE_TARGET)
        return _r


    def stop(self):
        self.teensy.write([self.CMD_STOP])
        _r = self.read(self.CMD_STOP)

    def read(self, cmd_byte):
        _b = self.teensy.read(1)
        if not _b:
            return
        while _b[0] != cmd_byte | 0x80:
            if _b[0] == self.REPLY_LOG:
                _len = int.from_bytes(self.teensy.read(2), 'big')
                _msg = self.teensy.read(_len)
                if self.log:
                    logging.debug(_msg)
            _b = self.teensy.read(1)
        _b += self.teensy.read(1) # Return code
        _len = int.from_bytes(self.teensy.read(2), "big")
        _b += self.teensy.read(_len)
        return _b[1:] # strip off the command response byte


if __name__ == "__main__":
    logging.basicConfig(level=logging.DEBUG, format="%(filename)s:%(funcName)s: %(message)s")
    cs = Teensy_CS(log = 1)
    cs.set_pulse_width(100)
    cs.set_delay(4500000) # start time
    delay_start =  100000
    for delay in range(delay_start, 5000000, 1000):
        cs.set_delay(delay) # start time
        cs.setup_target()
        _r = cs.run(1)
        print(_r)
        _r = cs.check()
        print(_r)

File: Teensy_Cs.cpp
#include "Teensy_Cs.h"


char log_buf[0x1000];



Teensy_Cs::Teensy_Cs(int trig_out, Cs_Target *target){
	pinMode(trig_out, OUTPUT);
	digitalWriteFast(trig_out, LOW);
	this->trig_out = trig_out;
	this->target = target;
}


void Teensy_Cs::set_trig_in(int trig_in, bool trig_in_level, unsigned long trig_in_delay_muS){
	this->trig_in = trig_in;
	this->trig_in_level = trig_in_level;
	this->trig_in_delay_muS = trig_in_delay_muS;
	pinMode(trig_in, INPUT);

	this->tr_in = 1;
}

int Teensy_Cs::trigger(){
	unsigned int delayMus = (unsigned int) (this->delayNs / 1000);
	unsigned long time = micros();
	// If trig_in pin is set, wait for this pin to go to the required level
	if(this->tr_in){
		// If we have an analog trigger_in pin
		if(this->tr_in == 2){
			/*
			int flag = 0;
			while(1){
				sprintf(log_buf, "Analog read %d, ref analog read: %d", this->adc.analogRead(this->trig_in), this->analog_trig_value);
				log(log_buf, strlen(log_buf));
				if(this->adc.analogRead(this->trig_in) < 220)
					flag = 1;
				if(flag && this->adc.analogRead(this->trig_in) == 255)
					return ERR_MCU_TRIG_IN;
			}
			*/
			while(this->adc.analogRead(this->trig_in) > 208); // Wait for trigger to reset - TODO not sure if problem with other trigger setups + can lead to infinite loop
			while(this->adc.analogRead(this->trig_in) < this->analog_trig_value){
				//sprintf(log_buf, "Analog read %d, ref analog read: %d", this->adc.analogRead(this->trig_in), this->analog_trig_value);
				//log(log_buf, strlen(log_buf));
				if(micros() > time + this->trig_in_delay_muS)
					return ERR_MCU_TRIG_IN;
			}

		}
		else if(this->tr_in == 1){
			while(digitalReadFast(this->trig_in) == this->trig_in_level); // Wait for trigger to reset - TODO not sure if problem with other trigger setups
			while(digitalReadFast(this->trig_in) != this->trig_in_level){
				if(micros() > time + this->trig_in_delay_muS)
					return ERR_MCU_TRIG_IN;
			}
		}
	}
	sprintf(log_buf, "Triggering %d", this->trig_out);
	log(log_buf, strlen(log_buf));
	delayMicroseconds(delayMus);
	DELAY(this->delayNs % 1000);
	if(trig_out_pulse == 1){
		digitalWriteFast(this->trig_out, HIGH);
		DELAY(this->widthNs);
		digitalWriteFast(this->trig_out, LOW);
	}
	else{
		digitalWriteFast(this->trig_out, LOW);
		DELAY(this->widthNs);
		digitalWriteFast(this->trig_out, HIGH);
	}
	return 0;
}

int Teensy_Cs::run_target(bool trigger, uint8_t* out, uint16_t* out_len){
	int res;
	res = this->target->pre_glitch(out, out_len);
	if(res != 0)
		return res;
	if(trigger)
		res = this->trigger();
	if(res != 0)
		return res;
	res = this->target->post_glitch(out, out_len);
	return res;
}


void Teensy_Cs::run(){
	uint16_t ret_len = 0;
	int ret = 0;
	int cmd = 0;
	char in[0x100] = {0};
	uint8_t ret_data[0x2000] = {0};
	this->target->setup();
	while(1){
		while(!Serial);
		cmd = Serial.read();
		if(cmd != -1){
			ret_len = 0;
			memset(ret_data, 0, 0x20);
			switch(cmd){
				case CMD_SET_WIDTH:
					Serial.readBytes(in, 4); // New width sent in little endian order
					this->set_widthNs(*((int*) in));
					sprintf(log_buf, "Setting width %d", this->widthNs);
					log(log_buf, strlen(log_buf));
					break;
				case CMD_SET_DELAY:
					Serial.readBytes(in, 4); // New width sent in little endian order
					this->set_delayNs(*((int*) in));
					sprintf(log_buf, "Setting delay %d", this->delayNs);
					log(log_buf, strlen(log_buf));
					break;
				case CMD_RUN:
					ret = this->run_target(0, ret_data, &ret_len);
					break;
				case CMD_CHECK:
					this->target->check(ret_data, &ret_len);
					break;
				case CMD_RUN_TRIG:
					ret = this->run_target(1, ret_data, &ret_len);
					break;
				case CMD_STOP:
					return;
				case CMD_PREPARE_MCU:
					ret = this->target->prepare_target();
					break;
				default:
					sprintf(log_buf, "[%02X] unknown command byte", cmd);
					log(log_buf, strlen(log_buf));
			}
			Serial.write(cmd | 0x80);
			Serial.write(ret);
			Serial.write(ret_len >> 8);
			Serial.write(ret_len & 0xff);
			for(int i = 0; i < ret_len; ++i)
				Serial.write(ret_data[i]);
		}
	}
}


void log(const char* log_str, uint16_t len){
	Serial.write(REPLY_LOG);
	Serial.write(len >> 0x8);
	Serial.write(len & 0xff);
	Serial.write(log_str, len);
	memset(log_buf, 0, len);
}

